### PR TITLE
ble: profile rework

### PR DESCRIPTION
**Pull request by hedger**
# What's new

**General BLE**:
- ble: reworked profile subsystem
- now applications can define custom BLE profiles
- profiles define BLE advertising parameters and own service set

**BLE internals**:
- combined SHCI & HCI event processing threads
- reworked event dispatching to services to support new way of defining profiles and overcome fixed wpan handler count limit
- battery service (&others): now supports multiple instances
- bt service: now only init BLE pairing dialog when needed to save memory
- moved BLE HID profile from furi_hal to lib/ble_profile as an example (custom profiles can also be defined in app's private libs)

**Extra beacon**:
- added API for controlling additional, non-connectable BLE advertisements (BLE beacon)
- that beacon works in background, independently from main BLE subsystem
- extra beacon has status indication on desktop
- added example app for configuring and testing extra beacon feature

**Misc**:
- Moved example_custom_font app from debug to examples
- Improved "ps" CLI command output: added thread priorities
- Added FURI_PACKED macro for packed structures
- hid_app: cleaner code split for USB/BLE transports
- fbt: better handling of malformed target.json files

# Verification 

- test that update from older versions and downgrade do not break BLE pairing
- run BLE remote app, check that old pairings work & app itself works
- run BLE stress test from companion apps
- run BLE Beacon app in various combinations with BLE on/off, RPC active/not
- check out cool "ps"
- idk

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: applications/examples/example_ble_beacon/ble_beacon_app.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+#include <rpc/rpc_app.h>
+#include <notification/notification_messages.h>
+
+#include <furi_hal_ble.h>
+
+#include "scenes/scenes.h"
+#include <stdint.h>
+
+#define TEXT_STORE_SIZE 64U
```
Comment (skotopes) [inline applications/examples/example_ble_beacon/ble_beacon_app.h]: remove

File: applications/main/nfc/plugins/supported_cards/opal.c [hunk 2]
```diff
@@ -28,6 +28,7 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include "core/common_defines.h"
```
Comment (skotopes) [inline applications/main/nfc/plugins/supported_cards/opal.c]: `<>`

File: applications/services/bt/bt_service/bt.h [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 #include <stdint.h>
 #include <stdbool.h>
+#include <furi_ble/profile_interface.h>
+#include <core/common_defines.h>
```
Comment (skotopes) [inline applications/services/bt/bt_service/bt.h:6]: remove?

File: applications/services/cli/cli_commands.c [hunk 4]
```diff
@@ -383,27 +384,33 @@ void cli_command_ps(Cli* cli, FuriString* args, void* context) {
 
     const uint8_t threads_num_max = 32;
     FuriThreadId threads_ids[threads_num_max];
-    uint8_t thread_num = furi_thread_enumerate(threads_ids, threads_num_max);
+    uint32_t run_times[threads_num_max];
+    uint32_t thread_num = furi_thread_enumerate(threads_ids, run_times, threads_num_max);
```
Comment (skotopes) [inline applications/services/cli/cli_commands.c]: Use DWT as time source
Snapshot values on start
Wait for 1 second
Get another values and calculate diff
Show 1 second usage


or make TOP command where you do it in cycle + sort processes by id/name

File: lib/ble_profile/extra_profiles/hid_profile.c [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+    // Change name Flipper -> Control
+    const char* clicker_str = "Control";
+    memcpy(&config->adv_name[1], clicker_str, strlen(clicker_str));
+}
+
+static const FuriHalBleProfileConfig profile_callbacks = {
+    .start = ble_profile_hid_start,
+    .stop = ble_profile_hid_stop,
+    .get_gap_config = ble_profile_hid_get_config,
+    // .process = NULL,
```
Comment (skotopes) [inline lib/ble_profile/extra_profiles/hid_profile.c]: remove

File: targets/f7/ble_glue/ble_system.c [hunk 6]
```diff
@@ -479,3 +428,28 @@ BleGlueCommandResult ble_glue_fus_wait_operation() {
...
+    if(info->magic != BLE_SYSTEM_HARDFAULT_INFO_MAGIC) {
+        return NULL;
+    }
+    return info;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// AN5289, 4.9
+
+void shci_cmd_resp_wait(uint32_t timeout) {
```
Comment (skotopes) [inline targets/f7/ble_glue/ble_system.c]: nope, remove it)

File: targets/f7/ble_glue/furi_ble/event_dispatcher.c [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    return ack_status;
+}
+
+void ble_service_event_dispatcher_init() {
+    furi_assert(!initialized);
+
+    GapSvcEventHandlerList_init(handlers);
+    SVCCTL_RegisterSvcHandler(ble_service_event_dispatcher_dispatch_event);
```
Comment (skotopes) [inline targets/f7/ble_glue/furi_ble/event_dispatcher.c]: Override weak symbols like SVCCTL_UserEvtRx and serve it from your code

File: targets/furi_hal_include/furi_hal_ble.h [hunk 8]
```diff
@@ -0,0 +1,294 @@
+/**
+ * @file furi_hal_ble.h
+ * BT/BLE HAL API
+ */
+
+#pragma once
+
+#include "core/string.h"
```
Comment (skotopes) [inline targets/furi_hal_include/furi_hal_ble.h]: remove

File: targets/furi_hal_include/furi_hal_ble.h [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+#include <gap.h>
+#include <extra_beacon.h>
+#include <furi_ble/profile_interface.h>
+#include <ble_system.h>
+#include <ble_stack.h>
+#include <stdint.h>
+
+#define FURI_HAL_BLE_STACK_VERSION_MAJOR (1)
+#define FURI_HAL_BLE_STACK_VERSION_MINOR (12)
+#define FURI_HAL_BLE_C2_START_TIMEOUT 1000
```
Comment (skotopes) [inline targets/furi_hal_include/furi_hal_ble.h]: ciferki v skobochkah

**Review by skotopes — CHANGES_REQUESTED** [review]

**Comment by thejudge156** [comment]
@hedger Would you be open to a PR to your branch that adds a OpenHaystack profile?

**Comment by hedger** [comment]
@thejudge156 you can bring that profile along with your custom app. Just implement the profile interface.

Or, even better, use the BLE beacon API - then it will properly work in the background. It won't rotate the keys on time, though, until the app is re-launched.

**Comment by thejudge156** [comment]
Thanks for the pointer, I will work on that.

**Review by skotopes — APPROVED** [review]

**Pull request merged**